src/main/java/net/rubygrapefruit/platform/internal/jni/AbstractFileEventFunctions.java

```diff
@@ -46,25 +50,47 @@ public AbstractWatcherBuilder(FileWatcherCallback callback) {
      * Configures a new watcher using a builder. Call {@link AbstractWatcherBuilder#start()} to
      * actually start the {@link FileWatcher}.
      */
-    public abstract AbstractWatcherBuilder newWatcher(FileWatcherCallback callback);
+    public abstract AbstractWatcherBuilder newWatcher(BlockingQueue<FileWatchEvent> queue);
 
     protected static class NativeFileWatcherCallback {
-        private final FileWatcherCallback delegate;
 
-        public NativeFileWatcherCallback(FileWatcherCallback delegate) {
-            this.delegate = delegate;
+        private final BlockingQueue<FileWatchEvent> eventQueue;
+
+        public NativeFileWatcherCallback(BlockingQueue<FileWatchEvent> eventQueue) {
+            this.eventQueue = eventQueue;
         }
 
         // Called from the native side
         @SuppressWarnings("unused")
-        public void pathChanged(int type, String path) {
-            delegate.pathChanged(FileWatcherCallback.Type.values()[type], path);
+        public void pathChanged(int typeIndex, String path) throws InterruptedException {
+            FileWatchEvent.Type type = FileWatchEvent.Type.values()[typeIndex];
+            if (type == FileWatchEvent.Type.OVERFLOWED) {
+                signalOverflow(path);
+            } else {
+                queueEvent(new ChangeEvent(type, path), false);
+            }
         }
 
         // Called from the native side
         @SuppressWarnings("unused")
-        public void reportError(Throwable ex) {
-            delegate.reportError(ex);
+        public void reportError(Throwable ex) throws InterruptedException {
+            queueEvent(new ErrorEvent(ex), true);
+        }
+
+        private void queueEvent(FileWatchEvent event, boolean deliverOnOverflow) throws InterruptedException {
+            // TODO Make the timeout configurable
+            if (!eventQueue.offer(event, 1, SECONDS)) {
```

Comment:
I am not sure if waiting here makes sense. If the blocking queue is full, then we have an overflow - let's clear it and add the overflow event. That also has the advantage that there won't be an interrupted exception.

Reply:
Alright, you convinced me. I'll remove this wait.

Reply:
I'll add it on top of #185.

Reply:
Actually, I won't. Let's finish this here in this PR, and I'll rebase #185 later.